keps/sig-storage/3751-volume-attributes-class/README.md

```diff
@@ -0,0 +1,1279 @@
+
+
+```
+vacInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+        AddFunc: e.vacAddedUpdated,
+        UpdateFunc: func(old, new interface{}) {
+                e.vacAddedUpdated(new)
+        },
+})
+
+pvcInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+        AddFunc: func(obj interface{}) {
+                e.pvcAddedDeletedUpdated(nil, obj, false)
+        },
+        DeleteFunc: func(obj interface{}) {
+                e.pvcAddedDeletedUpdated(nil, obj, true)
+        },
+        UpdateFunc: func(old, new interface{}) {
+                e.pvcAddedDeletedUpdated(old, new, false)
+        },
+})
+
+```
+
+After a delete VolumeAttributesClass request is made, it will add a deletionTimestamp and then confirm with the vac_protection_controller to see if there are any finalizers of that VolumeAttributesClass. The VolumeAttributesClass will only be deleted if there are no more finalizers.
+
+If the user decides to create a new VolumeAttributesClass with the exact same name, the creation will fail because the Kubernetes API server will refuse creating new objects with the same name until the old objects are deleted.
+
+![Delete VolumeAttributesClass Flow](./Delete_VolumeAttributesClass.png)
+
+#### Create PVC and Create Volume
+
+The pvc config will have a new field VolumeAttributesClassName.
+
+A new finalizer **kubernetes.io/vac-protection** will be added to the VolumeAttributesClass the PVC is consuming from a new vacProtectionController. We also need a new admission controller [vacobjectinuseprotection](https://github.com/kubernetes/kubernetes/tree/master/plugin/pkg/admission/storage)/admission.go.
+
+![Create PVC Flow](./Create_PVC.png)
+
+The components involved in CreateVolume:
+
+![Create Volume Flow](./Create_Volume.png)
+
+#### Delete PVC
+
+Deleting a PVC will trigger a list PVCs call and decide if we need to remove the finalizer **[kubernetes.io/vac-protection](kubernetes.io/vpc-protection) **via the vac_protection_controller.
+
+![Delete PVC Flow](./Delete_PVC.png)
+
+#### Modify PVC
+
+![VolumeAttributesClass Update Flow](./VolumeAttributesClass-Flow.png)
+
+Since VolumeAttributesClass is **immutable**, to update the parameters, the end user can modify the PVC object to set a different VolumeAttributesClass. If the existing VolumeAttributesClass cannot satisfy the end user’s use case, the end user needs to contact the cluster administrator to create a new VolumeAttributesClass.
+
+**Watching changes in the PVC object**, if the PVC’s VolumeAttributesClass changes, it will trigger a ModifyVolume call.
```

Comment:
Also ModifyVolume should only be allowed on bound PVC's.

Reply:
Added to line 580